rfcs/xxxx-authorization/text.md

```diff
@@ -0,0 +1,196 @@
+
+This discussion assumes that the obfuscated path has this form:  `/e/<obfuscated_uuid>/<datset_name>
+.cxg`, although other options are possible.
+
+In this way, it is easy to separate out the two components from the path.
+
+For obfuscated datasets, the authorization portion will go through an API to check the
+obfuscated_uuid and dataset.  It will ensure that the obfuscated_uuid belongs to a collection
+that contains the dataset, and it will ensure that the obfuscated_uuid has not expired.
+  
+`GET /v1/dataset/access_obfuscated`
+
+**Description:** determine if the obfuscated_uuid provides access to the dataset
+
+**Parameters:**
+
+1. obfuscated_uuid [REQUIRED]:  the collection's obfuscated_uuid
+2. dataset_name [REQUIRED]: the name of the dataset (e.g. "pbmc3k.cxg")
+
+**Response:**
+
+1. 200 OK, if the obfuscated_uuid is valid (and not expired) obfuscated_uuid for the
+ collection containing the dataset
+2. 401 Unauthorized
+3. 404 dataset not found or obfuscated_uuid not found.
+
+#### Community releaseas vs hosted cellxgene.
+
+There should be a configuration option to set the authorization type, such as how we do with
+authentication.  The authorization type we use for CZI hosted cellxgene possibly may be better
+implemented as a plugin, using our plugin architecture, so as not to make public the
+details of the implementation.
+
+In the scenario described in this document, the authorization component will need to answer the
+question: does the current user have access to the dataset at this path?  It will also need to
+return the dataset name from the path, since in the case of the obfuscated_uuid feature, the name 
+is embedded in the path, and needs to be extracted.  
+ 
+#### Caching issues
+
+Checking authorization requires access to the cellxgene explorer backend, which accesses the
+cellxgene portal backend.
+ 
+I need to do some research to determine how to force the authorization check on each API request
+to the explorer, but also caches copies of the data on cloudfront.
+
+### cellxgene Portal
+
+The `/v1/dataset/access/` and `v1/dataset/access_obfuscated` APIs will be implemented in the portal
+ code base, since they will need access to the portal database.
+
+To implement this efficiently, it may be useful to create two database views: One that maps
+`dataset.name` to `collection.owner` and `collection.visibility`.  And one that maps
+ `obfuscated_uuid` and `dataset.name` to `collection.obfuscated_uuid_expiration`.
+ 
+Note, `obfuscated_uuid_expiration` is not currently part of the database.
+
+
+### UX issues
+
+#### portal
+
+1. Can the owner create new obfuscated_uuid's, and set an expiration?
+1. The owner will need to retrieve the obfuscated URL of the collection, so it can be shared.
+1. Can there be more than one obfuscated_uuid?
+1. The owner may need to revoke a obfuscated_uuid, will that be possible?
+
+#### explorer
+
+1. What should be displayed if the user is not authorized to see a dataset? 
```

Comment:
Under what conditions? For example, a reviewer who is provided with a bearer token is not required to authenticate. If a non-reviewer guesses the capability url, then they would have access.

Reply:
This is what I had in mind.  Say a user goes to  ".../e/dataset.cxg", but that dataset is private and they are not the owner.    (Maybe they are the owner, but not logged into right now, so it is reasonable they have the URL).  All the explorer backend API's would return 401 errors in this case.   The explorer frontend, by default, would just show "Error" in red.   But we may want to say something nicer like "Sorry, you don't have access to this dataset".     A similar situation could arise for revoked capability urls.

Reply:
Why would the FE show _"Error" in red_ rather than prompting the user to authenticate in that case? 

You may want to validate with Trust/Infra but sometimes there's a recommendation to simply return a 404 for security reasons because a 401 is leaking hints to potential attackers.  For example, see the [comments](https://security.stackexchange.com/questions/150758/returning-the-wrong-http-response-code-on-purpose).

And the behavior is supported in [RFC7231](https://tools.ietf.org/html/rfc7231#section-6.5.4)

_The 404 (Not Found) status code indicates that the origin server did not find a current representation for the target resource or **is not willing to disclose that one exists**._

Although it's confusing or [annoying](https://github.com/dear-github/dear-github/issues/162) to some users in practice.

What about the case where I've authenticated and then paste the url that I'm not authorized to use. Does the system return a 403? For example, see this[ thread](https://lists.w3.org/Archives/Public/ietf-http-wg/2018AprJun/0314.html).

Reply:
I'm saying that by default, right now, in the explorer, if the /config or other APIs return and error code, then it will be handled ungracefully (red error boxes).   So the point of this bullet was to raise the issue so that we could decide what we want to show instead.

Reply:
I've reworded this part in the document to make it less confusing (I hope)